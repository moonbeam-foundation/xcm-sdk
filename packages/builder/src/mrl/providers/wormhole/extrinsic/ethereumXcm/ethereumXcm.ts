import { type Address, encodeFunctionData } from 'viem';
import { ERC20_ABI } from '../../../../../balance/Erc20Abi';
import type { ContractConfig } from '../../../../../types/evm/ContractConfig';
import { ExtrinsicConfig } from '../../../../../types/substrate/ExtrinsicConfig';
import { BATCH_CONTRACT_ADDRESS } from '../../../../MrlBuilder.constants';
import type { MrlConfigBuilder } from '../../../../MrlBuilder.interfaces';
import { contract as ContractBuilder } from '../../contract';
import { BATCH_CONTRACT_ABI } from './BatchContractAbi';

export function ethereumXcm() {
  return {
    transact: (): MrlConfigBuilder => ({
      build: (params) => {
        const { asset, isAutomatic, moonChain, moonGasLimit } = params;

        if (!moonGasLimit) {
          throw new Error('moonGasLimit must be defined');
        }

        const tokenAddressOnMoonChain = moonChain.getChainAsset(asset)
          .address as Address | undefined;

        if (!tokenAddressOnMoonChain) {
          throw new Error(
            `Asset ${asset.symbol} does not have a token address on chain ${moonChain.name}`,
          );
        }

        const tokenAmountOnMoonChain = asset.convertDecimals(
          moonChain.getChainAsset(asset).decimals,
        ).amount;

        const contract = (
          isAutomatic
            ? ContractBuilder()
                .TokenBridgeRelayer()
                .transferTokensWithRelay()
                .build(params)
            : ContractBuilder().TokenBridge().transferTokens().build(params)
        ) as ContractConfig;

        const approveTx = encodeFunctionData({
          abi: ERC20_ABI,
          functionName: 'approve',
          args: [contract.address as Address, tokenAmountOnMoonChain],
        });

        const batchAll = encodeFunctionData({
          abi: BATCH_CONTRACT_ABI,
          functionName: 'batchAll',
          args: [
            [tokenAddressOnMoonChain, contract.address as Address],
            [0n, 0n], // Value to send for each call
            [approveTx, contract.encodeFunctionData()], // Call data for each call
            [], // Gas limit for each call
          ],
        });

        return new ExtrinsicConfig({
          module: 'ethereumXcm',
          func: 'transact',
          getArgs: () => [
            {
              V2: {
                gasLimit: moonGasLimit,
                action: {
                  Call: BATCH_CONTRACT_ADDRESS,
                },
                value: 0,
                input: batchAll,
              },
            },
          ],
        });
      },
    }),
  };
}
