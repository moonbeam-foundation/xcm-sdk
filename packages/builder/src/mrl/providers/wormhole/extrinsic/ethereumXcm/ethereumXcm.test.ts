import { describe, expect, it } from 'vitest';

import {
  mrlBuildParamsMock,
  mrlBuildParamsMock2,
} from '../../../../../../fixtures';
import type { ExtrinsicConfig } from '../../../../../extrinsic';
import { ethereumXcm } from './ethereumXcm';

describe('ethereumXcm', () => {
  describe('transact', () => {
    const extrinsic = ethereumXcm()
      .transact()
      .build(mrlBuildParamsMock) as ExtrinsicConfig;

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });

    it('should throw and error because destination has no wh name', () => {
      expect(() =>
        (
          ethereumXcm().transact().build(mrlBuildParamsMock2) as ExtrinsicConfig
        ).getArgs({} as any),
      ).toThrow('Chain Interlay Testnet does not have a wormhole name');
    });
  });
});
