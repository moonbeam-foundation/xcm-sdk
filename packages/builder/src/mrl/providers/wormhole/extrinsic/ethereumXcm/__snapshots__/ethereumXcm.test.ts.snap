// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ethereumXcm > transact > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transact",
  "getArgs": [Function],
  "module": "ethereumXcm",
}
`;

exports[`ethereumXcm > transact > should get correct arguments 1`] = `
[
  {
    "V2": {
      "action": {
        "Call": "0x0000000000000000000000000000000000000808",
      },
      "gasLimit": 999999n,
      "input": "0x96e292b8000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000320000000000000000000000000000000000000000000000000000000000000000200000000000000000000000098891e5fd24ef33a488a47101f65d212ff6e650e0000000000000000000000009563a59c15842a6f322b10f69d1dd88b41f2e97b0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000044095ea7b30000000000000000000000009563a59c15842a6f322b10f69d1dd88b41f2e97b000000000000000000000000000000000000000000000000000000170cdc1e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c41019d65400000000000000000000000098891e5fd24ef33a488a47101f65d212ff6e650e000000000000000000000000000000000000000000000000000000170cdc1e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000eF46c7649270C912704fB09B75097f6E32208b850000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "value": 0,
    },
  },
]
`;
